### PR TITLE
add release ratio to meta field on the end_filter tracer method

### PR DESCRIPTION
**Pull request by lucila**
### What and why? 🤔

The `end_filter` event will now log `released_ratio` as part of the meta fields. The released rate is computed as : `number of released elements /  total number of elements`. 

If the filter step is called with an empty array, the release count and release rate will be 0.  And "release_all" will not be logged as event.

### Testing Steps ✍️

Code review. Unit tests should all pass.

### You're it! 👑

@Automattic/stream-builders

### CODE REVIEW

File: tests/unit/Tumblr/StreamBuilder/StreamFilters/StreamFilterTest.php [hunk 1]
```diff
@@ -58,15 +57,15 @@ public function test_filter()
         // Example output:
         // [2024-01-16T09:28:33-05:00]: op=filter sender=ello[Mock_StreamFilter_e7231ac5] status=begin other={"count":2}
         // [2024-01-16T09:56:32-05:00]: op=filter sender=ello[Mock_StreamFilter_3aaefcd3] status=end
-        //      start_time=1705416992.4759 duration=2.0980834960938E-5 other={"count":1}
+        //      start_time=1705427127.9359 duration=7.1525573730469E-6 other={"count":2,"released_ratio":1}
```
Comment (lucila) [inline tests/unit/Tumblr/StreamBuilder/StreamFilters/StreamFilterTest.php]:
```suggestion
        //      start_time=1705427127.9359 duration=7.1525573730469E-6 other={"count":1,"released_ratio":1}
```

File: lib/Tumblr/StreamBuilder/StreamFilters/StreamFilter.php [hunk 2]
```diff
@@ -133,7 +134,7 @@ final public function filter(
     /**
      * The filter concrete logic implementation.
      * @param StreamElement[] $elements The candidate StreamElement instances to filter.
-     * @param StreamFilterState $state The filter state passed from previous filter operations.
+     * @param StreamFilterState|null $state The filter state passed from previous filter operations.
```
Comment (cyle) [inline lib/Tumblr/StreamBuilder/StreamFilters/StreamFilter.php:137]: i think the argument itself needs to be updated to `?StreamFilterState` for this to be true....

**Comment by nicola-barbieri** [comment]
> If the filter step is called with an empty array, the release count and release rate will be 0

i think we should just skip the logging in that case

File: lib/Tumblr/StreamBuilder/StreamTracers/StreamTracer.php [hunk 3]
```diff
@@ -387,21 +387,30 @@ final public function begin_filter(StreamFilter $filter, int $in_count): void
...
         int $released_count,
-        array $timing
+        array $timing,
+        int $total_count
     ): void {
+        $meta = [static::META_COUNT => $released_count];
+
+        if ($total_count > 0) {
+            $ratio = $released_count / $total_count;
+            $meta['released_ratio'] = $ratio;
```
Comment (nicola-barbieri) [inline lib/Tumblr/StreamBuilder/StreamTracers/StreamTracer.php]: should we apply some rounding to avoid too many digits ?

File: tests/unit/Tumblr/StreamBuilder/StreamFilters/StreamFilterTest.php [hunk 4]
```diff
@@ -45,28 +45,29 @@ public function test_filter()
...
         // [2024-01-16T09:56:32-05:00]: op=filter sender=ello[Mock_StreamFilter_3aaefcd3] status=release
         //      other={"target":"MockMaxStreamElement","meta_detail":"TEST_MockMaxElement(123)","filter_code":"Mock_StreamFilter_3aaefcd3"}
         $this->assertMatchesRegularExpression(
-            "/op=filter.sender=ello\[Mock_StreamFilter_[a-z0-9A-Z]*\].status=begin.other=\{\"count\":2\}/",
+            "/op=filter.sender=ello\[Mock_StreamFilter_[a-z0-9A-Z]*\].status=begin.other=\{\"count\":3\}/",
             $tracer->get_output()[0]
         );
         $this->assertMatchesRegularExpression(
-            "/op=filter.sender=ello\[Mock_StreamFilter_[a-z0-9A-Z]*\].status=end.start_time=.*duration=.*other=\{\"count\":1\}/",
+            "/op=filter.sender=ello\[Mock_StreamFilter_[a-z0-9A-Z]*\].status=end.start_time=.*duration=.*other=\{\"count\":1,\"released_ratio\":0.333\}/",
```
Comment (lucila) [inline tests/unit/Tumblr/StreamBuilder/StreamFilters/StreamFilterTest.php:70]: round released_ratio ⬆️  cc: @nicola-barbieri

**Comment by lucila** [comment]
> > If the filter step is called with an empty array, the release count and release rate will be 0
> 
> i think we should just skip the logging in that case

@nicola-barbieri 
When the array is empty, the `begin` and `end` ops are traced, with their times. 
I changed this so that if the array is empty, we trace the `skip` event and return immediately

**Review by cyle — APPROVED** [review]
groovy

File: lib/Tumblr/StreamBuilder/StreamFilters/StreamFilter.php [hunk 5]
```diff
@@ -133,11 +134,11 @@ final public function filter(
...
      * @param StreamTracer|null $tracer The tracer passed in to track filter behaviors.
      * @return StreamFilterResult
      */
-    abstract protected function filter_inner(array $elements, StreamFilterState $state = null, StreamTracer $tracer = null): StreamFilterResult;
+    abstract protected function filter_inner(array $elements, ?StreamFilterState $state = null, ?StreamTracer $tracer = null): StreamFilterResult;
```
Comment (nicola-barbieri) [inline lib/Tumblr/StreamBuilder/StreamFilters/StreamFilter.php:141]: why is this change needed ?

**Review by nicola-barbieri — APPROVED** [review]

File: lib/Tumblr/StreamBuilder/StreamFilters/StreamFilter.php (hunk 5 above)
Comment (lucila) [reply lib/Tumblr/StreamBuilder/StreamFilters/StreamFilter.php:141]: @nicola-barbieri this change is unrelated to the goal of this PR. but the php docs were not consistent with the function arguments and I was seeing a warning on my local env. I fixed it since I was already modifying this class.
